### PR TITLE
BUG: ensure that docstrings for read/write public methods are always available for inspection (`nddata`)

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
ref #17472

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
The errors here are puzzling. I'll come back to this later today if I have time.

**Comment by neutrinoceros** [comment]
Actually the problem was quite trivial, I'm just paying the price of context switching.

File: astropy/io/registry/interface.py [hunk 1]
```diff
@@ -99,7 +99,6 @@ def help(self, format=None, out=None):
             reader_doc += os.linesep
             if doc is not None:
                 reader_doc += inspect.cleandoc(doc)
-
```
Comment (nstarman) [inline astropy/io/registry/interface.py:102]: ?

File: astropy/nddata/ccddata.py [hunk 2]
```diff
@@ -738,6 +682,66 @@ def fits_ccddata_reader(
     return ccd_data
 
 
+# this docstring *must* be available at runtime to support `UnifiedReadWrite.help`
+# so we attach it manually, which is robust against PYTHONOPTIMIZE=2
+fits_ccddata_reader.__doc__ = """
```
Comment (nstarman) [inline astropy/nddata/ccddata.py:687]: A quick google search didn't reveal anything about being able to make a decorator that prevents the docstring from being removed. However if the optimizer is just doing static AST parsing then perhaps a simple decorator would work.

```
@keep_doc("""docstring""")
def function()
```

Worth a try? Keeps the docs and function more closely associated.

Comment (neutrinoceros) [reply astropy/nddata/ccddata.py:687]: I don't have a strong opinion, but it's certainly feasible. I'll push an implementation of this in a minute, let me know what you think !

Comment (nstarman) [reply astropy/nddata/ccddata.py]: Does this works with the optimizer flag?

**Comment by pllim** [comment]
tl;dr -- Didn't we decide in another PR that docstring missing in optimized mode is okay? So why this special treatment here?

**Comment by nstarman** [comment]
The help function requires the docstring. We could either: have a broken function, refactor the help function to be a no-op in optimized mode, or keep the docstring.

**Comment by pllim** [comment]
> refactor the help function to be a no-op

That would be simpler, right, and localize this optimized mode stuff in one file instead of spread it out over different subpackages that hook into the unified I/O?

**Comment by nstarman** [comment]
I suppose the Q though is: when do people use python's optimized mode? If it's only to do production runs, then making `.help()` a no-op is fine. If it's also to use Astropy in an IDE, then...

**Comment by eerovaher** [comment]
In Python the natural way of documenting things is using docstrings. The fact that a few `help()` methods rely on docstrings and break if they are not available is therefore not surprising. If you want to be helpful for users that do manage to stumble upon this problem then the sensible thing to do would be to raise a `RuntimeError` in the `help()` with an informative message.

Independently from the previous point, if a user has explicitly told Python to remove the docstrings then any clever tricks that insert the docstrings anyways clearly go against the user's wishes.

File: astropy/nddata/ccddata.py (hunk 2 above)
Comment (neutrinoceros) [reply astropy/nddata/ccddata.py]: It does. I should have said that when I pushed, but I did test it.

**Comment by neutrinoceros** [comment]
Yes, making `help` functions either no-ops or runtime errors in optimized mode would actually be more sensible and easier to implement/maintain. I'm slightly leaning towards "noop" but I don't have strong opinion. Would that work for you guys ?

**Comment by mhvk** [comment]
:+1: to not sneakily trying to keep docstrings, and instead adjusting `help`. I'd lean to *not* adjusting `help` at all and just letting it fail, but am OK with a more informative `RuntimeError` too. I don't like a no-op.

**Comment by saimn** [comment]
Someone using PYTHONOPTIMIZE=2 should not expect things like `.help()` to work. This is not a normal mode for e.g. interactive work (the only benefit to gain a few kb by **removing docstrings**), so can we avoid adding tricks and made the code more complex than needed to support invalid usecases ? (Same for `config` #17548 and maybe others?) 
`.help()` can just return None or even raise an exception in that case (and we should not care about it and waste time on this, there is really no reason to use this mode unless you need to fit a ton of code on limited hardware).

**Comment by neutrinoceros** [comment]
Yes, in hindsight this seems like the correct approach. I just didn't know how much tinkering I'd need before I gave it a try, but I agree that it's too much. Let's just close this, as well as #17557 and revert #17548.
I'll open a new PR tomorrow with the more consensual and simple approach.
Thanks all !

**Comment by mhvk** [comment]
Thanks, @neutrinoceros -- Sometimes one needs to see the code to realize it is the wrong approach!
